File: fritzStreams/src/main/java/de/acepe/fritzstreams/ui/fragments/CacheFragment.java
package de.acepe.fritzstreams.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import de.acepe.fritzstreams.backend.DownloadInfo;
import de.acepe.fritzstreams.backend.DownloadServiceAdapter;
import de.acepe.fritzstreams.backend.StreamInfo;

import java.util.Calendar;
import java.util.HashMap;

import static de.acepe.fritzstreams.backend.StreamInfo.Stream.NIGHTFLIGHT;
import static de.acepe.fritzstreams.backend.StreamInfo.Stream.SOUNDGARDEN;

/**
 * A simple {@link Fragment} subclass.
 */
public class CacheFragment extends Fragment {

    private static final String TAG = "CacheFragment";

    private HashMap<Calendar, StreamInfo> mSoundgardenStreamsForDay;
    private HashMap<Calendar, StreamInfo> mNightflightStreamsForDay;
    private Context mContext;
    private DownloadServiceAdapter downloadServiceAdapter;
    private Calendar mDay;

    /**
     * This method will only be called once when the retained Fragment is first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.i(TAG, "Cache Fragement created");

        // Retain this fragment across configuration changes.
        setRetainInstance(true);

        mSoundgardenStreamsForDay = new HashMap<>();
        mNightflightStreamsForDay = new HashMap<>();

        downloadServiceAdapter = new DownloadServiceAdapter();
        downloadServiceAdapter.attach(mContext);
    }

    /**
     * Hold a reference to the parent Activity so we can report the task's current progress and results. The Android
     * framework will pass us a reference to the newly created Activity after each configuration change.
     */
    @Override
    public void onAttach(Context activity) {
        super.onAttach(activity);
        mContext = activity;
        if (downloadServiceAdapter != null) {
            downloadServiceAdapter.attach(mContext);
        }
    }

    /**
     * Set the callback to null so we don't accidentally leak the Activity instance.
     */
    @Override
    public void onDetach() {
        super.onDetach();
        downloadServiceAdapter.detach();
        mContext = null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        downloadServiceAdapter.detachFromService();
    }

    public void setmDay(Calendar day) {
        this.mDay = day;
    }

    public Calendar getDay() {
        return mDay;
    }

    public void addStream(StreamInfo streamInfo) {
        if (streamInfo.getStream() == SOUNDGARDEN) {
            mSoundgardenStreamsForDay.put(streamInfo.getDay(), streamInfo);
            Log.i(TAG, "Added " + streamInfo);
        }
        if (streamInfo.getStream() == NIGHTFLIGHT) {
            mNightflightStreamsForDay.put(streamInfo.getDay(), streamInfo);
            Log.i(TAG, "Added " + streamInfo);
        }
    }

    public StreamInfo getStream(StreamInfo.Stream stream, Calendar day) {
        HashMap<Calendar, StreamInfo> streamsForDay = stream == NIGHTFLIGHT
                ? mNightflightStreamsForDay
                : mSoundgardenStreamsForDay;
        return streamsForDay.get(day);
    }

    public void scheduleDownload(DownloadInfo downloadInfo) {
        downloadServiceAdapter.addDownload(downloadInfo);
    }

    public DownloadServiceAdapter getDownloadServiceAdapter() {
        return downloadServiceAdapter;
    }

}

File: fritzStreams/src/main/java/de/acepe/fritzstreams/ui/fragments/CacheFragment.kt
package de.acepe.fritzstreams.ui.fragments

import android.content.Context
import android.os.Bundle
import android.os.Environment
import android.support.v4.app.Fragment
import android.util.Log
import de.acepe.fritzstreams.backend.*
import java.util.*

private const val TAG = "CacheFragment"

class CacheFragment : Fragment(), StreamCache {

    override var day: Calendar? = null

    private val streamManager: StreamManager by lazy {
        StreamManager(context!!.getExternalFilesDir(Environment.DIRECTORY_MUSIC).getAbsolutePath())
    }
    val downloadServiceAdapter: DownloadServiceAdapter by lazy {
        DownloadServiceAdapter()
    }


    /*
     * This method will only be called once (but after onAttach) when the retained Fragment is first created.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.d(TAG, "Cache Fragement created")

        // Retain this fragment across configuration changes.
        retainInstance = true
    }

    /*
     * Hold a reference to the parent Activity so we can report the task's current progress and results. The Android
     * framework will pass us a reference to the newly created Activity after each configuration change.
     */
    override fun onAttach(activity: Context) {
        super.onAttach(activity)
        downloadServiceAdapter.attach(activity)
    }

    /*
     * Set the callback to null so we don't accidentally leak the Activity instance.
     */
    override fun onDetach() {
        super.onDetach()
        downloadServiceAdapter.detach()
    }

    override fun onDestroy() {
        super.onDestroy()
        downloadServiceAdapter.detachFromService()
    }

    override fun getStream(stream: Stream, day: Calendar): OnDemandStream {
        return streamManager.getOrCreateStream(stream, day)
    }

    override fun scheduleDownload(downloadInfo: DownloadInfo) {
        downloadServiceAdapter.addDownload(downloadInfo)
    }

}
